pkg/nmi/server/server.go

```diff
@@ -174,11 +175,23 @@ func (s *Server) hostHandler(logger *log.Entry, w http.ResponseWriter, r *http.R
 		http.Error(w, "missing 'podname' and 'podns' from request header", http.StatusBadRequest)
 		return
 	}
-	podIDs, err := listPodIDsWithRetry(r.Context(), s.KubeClient, logger, podns, podname, rqClientID, listPodIDsRetryAttempts)
+	podIDs, identityInCreatedStateFound, err := listPodIDsWithRetry(r.Context(), s.KubeClient, logger, podns, podname, rqClientID, listPodIDsRetryAttemptsForCreated, listPodIDsRetryAttemptsForAssigned)
 	if err != nil {
 		msg := fmt.Sprintf("no AzureAssignedIdentity found for pod:%s/%s in assigned state", podns, podname)
 		logger.Errorf("%s, %+v", msg, err)
-		http.Error(w, msg, http.StatusNotFound)
+		var statusCode int
+		// if at least an identity was found in created state then we return 404 which is a retriable error code
+		// in the go-autorest library. If the identity is in CREATED state then the identity is being processed in
+		// this sync cycle and should move to ASSIGNED state soon.
+		if identityInCreatedStateFound {
+			statusCode = http.StatusNotFound
+		} else {
+			// if no identity in at least CREATED state was found, then it means the identity creation is not part of the
+			// current ongoing sync cycle. So we return 403 which a non-retriable error code so we give mic enough time to
+			// finish current sync cycle and process identity in the next sync cycle.
```

Comment:
how is that we give enough time for mic to finish current cycle by passing back a non-retriable error ?

Reply:
It just means there is no aggressive retry from the go-autorest sdk. Instead of retrying immediately in ~2s which is configured in sdk, now the retry is solely based on the client implementation.